Feature/remove galois in pre type one two primes

BarinderBanwait opened the pull request:
This PR has implemented all the changes we have recently discussed via call/email; they are all contained in the `PreTypeOneTwoPrimes` part of the code:

1. The field `K` is no longer required to be Galois. The computation proceeds via the embeddings of `K` into its galois closure `Kgal`.
2. The emergency aux primes have been corrected. To keep these to an absolute minimum for now (for testing), the emergency aux primes corresponding to generators of the class group have only been added in the case where `K` contains an imaginary quadratic field. (I eventually want to do this more generally, but for testing the `use_PIL` code to be described below, it is better to not add potentially large primes if not strictly necessary.)
3. The epsilons are now computed with respect to `K` and not `Kgal`. However, if `K` is itself Galois, then we still do the "mod out by galois action" idea on the epsilons.
4. The divisibilities from units is computed as a first step (see line 898 in `isogeny_primes.py`).
5. The `filter_ABC_primes` method is now fixed, using the observation that if the epsilon has different integers in it, then `p` cannot be inert.
6. Most significantly, the "Lambda" idea has been implemented. This all takes place in the block of code after `if use_PIL and h_K > 1:` on line 927 (this idea doesn't bring us anything if the class group is trivial). `use_PIL` is now a flag the user can request at the command line.

Let me say a few things about how (6) works. Firstly, when computing the `C` integers, the frobenius polynomials used for each auxiliary prime are stored in a dictionary `frob_polys_dict`, which is then used in the `use_PIL` code. The first step of this new code extracts the roots of those frobenius polynomials and creates essentially the set you defined in Definition 4.7 of the overleaf (expect that I've raised everything to the 12 at this point). It would help if we didn't need to keep track of the field in which all of these values are contained in, and I tried using coercions into `QQbar` but that didn't work, so the `beta_data` dict keeps track of the field in which all of the betas are defined as well as the betas themselves; so the dict is the tuple `(the_split_field, the_betas)`.

After this, the basis of Lambda is computed, using the code you supplied on CoCalc. Since we've already by this point computed the diagonal entries of this basis, we only care about doing computations on basis elements which are not diagonal; this is the list `good_basis_elements`.

We then want to compute, for each good basis element, the cartesian product of the different beta sets required, as well as the alpha corresponding to this basis element. this is eventually done in the `pre_eps_data_blob`. This requires one to compute the compositum `big_compositum` of some of the `the_split_field` number fields computed in the `beta_data` dict. This is done in the method `get_compositum`. Eventually, `pre_eps_data_blob` returns the alpha, all the products of the betas, as well as maps from `Kgal` and `big_compositum` into the compositum of `big_compositum` with `Kgal`, referred to as `huge` in the naming of those maps.

Having made these data blobs, we are then able to compute the norm in Proposition 4.8 for each epsilon and basis element. Running lcms and gcds are set-up to compute the integers and gcd them with what we got from before; this happens on line 971.

The problem with all of this is that it doesn't currently work, hence this being only a draft PR. It seems to get stuck in computing these big compositum fields, which goes back to the `get_compositum` function on line 665. This seems to be a bottleneck to this approach.

## Do you know  a better way to do this? 

Or even a way to not have to keep track of the fields where these products of betas are living? As I wrote above I did hope there would be a lazy way of doing this with `QQbar`, but I couldn't get that to work.

For the record I was running this code from the command line: `sage isogeny_primes.py 'x^3 - x^2 + 5*x + 1' --norm_bound 10 --verbose --use_PIL`

koffie commented:
Nice! This seems like a huge leap forward. Didn't expect for all the points to be implemented this fast.

> 3. The epsilons are now computed with respect to `K` and not `Kgal`. However, if `K` is itself Galois, then we still do the "mod out by galois action" idea on the epsilons.

I've thought about this a bit, and to me it seems there is a good chance we can even mod out by the action of `Kgal / Q` in the case `K != Q`. I have to work out the details still. But I think this could prove to be a worthwhile approach. Made #22 for this.

> 5. The `filter_ABC_primes` method is now fixed, using the observation that if the epsilon has different integers in it, then `p` cannot be inert.

Nice! In the case we have degree > 2 the criterion can actually be made a bit more generic (see definition 4.3 and corollary 4.4). Made issue #21 for this. So we can make this pr correct and mergeable without having to worry about implementing every improvement we have come up with

> 6. Most significantly, the "Lambda" idea has been implemented. This all takes place in the block of code after `if use_PIL and h_K > 1:` on line 927 (this idea doesn't bring us anything if the class group is trivial). `use_PIL` is now a flag the user can request at the command line.
> 
> The problem with all of this is that it doesn't currently work, hence this being only a draft PR. It seems to get stuck in computing these big compositum fields, which goes back to the `get_compositum` function on line 665. This seems to be a bottleneck to this approach.
> 
> ## Do you know a better way to do this?
> Or even a way to not have to keep track of the fields where these products of betas are living? As I wrote above I did hope there would be a lazy way of doing this with `QQbar`, but I couldn't get that to work.
> 
> For the record I was running this code from the command line: `sage isogeny_primes.py 'x^3 - x^2 + 5*x + 1' --norm_bound 10 --verbose --use_PIL`

Good to have a draft implementation that is highlighting computational difficulties with this approach! I'll have to think about how to best do it in practice. Maybe it is a good idea to leave the Lambda idea out of this pr for the moment so that we can already have the other useful features. And then work on the practical details of the Lambda implementation later?

p.s. I still have to review the code. So my response here are just some preliminary remarks on the text you wrote. I hope to make time for this the coming week. But if that doesn't work, I guess we could go through it our next meeting!

BarinderBanwait commented:
> Good to have a draft implementation that is highlighting computational difficulties with this approach! I'll have to think about how to best do it in practice.

Another idea I had meanwhile was maybe to not do things exactly, and instead use floating point arithmetic, and then use LLL later to reconstruct algebraic integers? We would then take the norm of this reconstructed integer just like in line 967. Since we eventually have to take a norm, I don't know if there's a way to get the rational integer entirely with approximation+LLL.

>  Maybe it is a good idea to leave the Lambda idea out of this pr for the moment so that we can already have the other useful features. And then work on the practical details of the Lambda implementation later?

Well most of the lambda idea stuff takes place with the `use_PIL` flag, so imo there's no harm done (since we're the only two clients of the code at this point) if that gets merged even in its currently 'in-development' state, as long as we remember not to call the code with this flag. However I agree that removing it from the (currently fairly large) PR makes it easier to review the resulting smaller PR, so perhaps you're right.

koffie commented on isogeny_primes.py:
```diff
@@ -730,85 +781,140 @@ def get_C_primes(K, G_K, frak_q, epsilons, q_class_group_order, loop_curves=Fals
-        AB_primes_dict = get_AB_primes(G_K,q,epsilons, q_class_group_order)
-        C_primes_dict = get_C_primes(Kgal,G_K, q, epsilons, q_class_group_order, loop_curves)
+        AB_integers_dict = get_AB_integers(embeddings,q,epsilons, q_class_group_order)
+        C_integers_dict, frob_polys = get_C_integers(Kgal,embeddings, q, epsilons, q_class_group_order, loop_curves)
```
I think it makes no sense to get the `frob_polys` from the `get_C_integers` function. The `get_weil_polys` function is ridiculously fast so it is something that we can just compute again on the fly. The bottleneck will be in other parts of the code. Especially since this ties the `get_C_integers` functionality together with the PIL code in an unatural way and make the code less maintainable.

koffie replied:
If you really want the `frob_polys_dict` (for caching and efficiency reasons) then my suggestion is that you compute it outside of the `get_C_integers` function using a seperate function whose sole purpose is to compute this dict. And give it as an (optional?) input to `get_C_integers`.

BarinderBanwait replied:
Good point -- the next commit has gone for the second of these suggestions, to compute the polys separately and pass the list to `get_C_integers`.

koffie commented on isogeny_primes.py:
```diff
@@ -818,6 +924,52 @@ def get_pre_type_one_two_primes(K, norm_bound=50, loop_curves=False):
         q_dict_collapsed = gcd(list(q_dict.values()))
         tracking_dict_inv_collapsed[eps] = q_dict_collapsed
 
+    if use_PIL and h_K > 1:
```
The `get_pre_type_one_two_primes` is getting a bit long and unwieldy. I think it is good to split it up into multiple functions. I think it makes sense to put some of the `use_PIL` stuff in a separate function that is called `get_PIL_integers` in the same spirit as the `get_C_integers`, and then have here only the code that mergers the results of the get_PIL_integers` with the other integers.

BarinderBanwait replied:
Good point -- next commit will fix this.

BarinderBanwait replied:
I've also taken the opportunity to split other stuff from this method into separate functions.

koffie commented on isogeny_primes.py:
```diff
@@ -730,85 +781,140 @@ def get_C_primes(K, G_K, frak_q, epsilons, q_class_group_order, loop_curves=Fals
+
+    # Do the unit computation first
+
+    unit_gens = K.unit_group().gens_values()
```
Maybe give the integers coming from units a name as well (get_U_integers) and put the code for it in a separate function.

BarinderBanwait replied:
done.

koffie reviewed (changes requested):
See the comments for requested changes

koffie commented:
> Well most of the lambda idea stuff takes place with the `use_PIL` flag, so imo there's no harm done (since we're the only two clients of the code at this point) if that gets merged even in its currently 'in-development' state, as long as we remember not to call the code with this flag.

Well ok this argument got me convinced to leave it in this PR but leave the code unreasonably slow. I will review this PR without paying attention to the speed of the code in the `use_PIL` flag being set, but I will give comments on the correctness and maintainability part of the `use_PIL` code. So that at least, if it terminates it will produce sensible results.

BarinderBanwait commented:
Addressed comments related to this PR in the latest commit.

BarinderBanwait commented:
The latest commit implements the suggested improvement in #23 using tensor products of companion matrices. I'm pleased to say that it does eventually finish on the running example:

```
sudo sage isogeny_primes.py 'x^3 - x^2 + 5*x + 1' --norm_bound 10 --verbose --use_PIL
```

Running the PIL code here took 6m20s; in this case the PIL code didn't sieve any pretender primes out, because even without the PIL code the set of PreTypeOneTwoPrimes were just the primes [2,3,5,7,11,17].

I didn't put in the code checking for perfect power when every PIL int is computed, since we end up taking gcds of these integers, so the right time to check if the integer is a perfect power (and return the base number if it is) is just before calling `prime_divisors`, which has been added into this PR.

BarinderBanwait commented:
Hi @koffie I'm going ahead and merging this into master now, since I've addressed all of your comments above, and before starting work on the filtering I'd like master to be up to date. In case you have any concerns or questions about the code in this PR you can raise a subsequent issue about that.
